### ISSUE TITLE
Safe Persistent Caching & Containerized Actions

### ISSUE THREAD

**Issue by nordlow** [mesonbuild/meson#1493]
Are there any plans on adding safe Bazel/Tup-like isolated/jailed execution (and in turn provably correct memoization) of actions in Meson?

We're currently doing of a build system survey at work and Bazel has come up as the most promising contender because of this, currently, unique feature.

**Comment by jpakkane** [comment]
Could you explain in a bit more detail what you specifically mean by this, especially what is the exact problem you are trying to solve.

Thanks,

**Comment by nordlow** [comment]
The problem of providing persistent (shared) caching of the execution of a build rule is similar to tracking exactly what is input and what is output in a calculation.

If you can restrict/forbid a build action to

- only execute a specific program (say gcc or ld),
- only access inputs from a, say, `in` directory tree (containing only the input files of the rule) and
- only write it's outputs to an `out` directory tree and 

then you can automatically provide automatic safe memoization/caching of that calculation. This is fundementally new paradigm that Tup and now also Bazel is striving for utilizing.

**Comment by nirbheek** [comment]
Is this something you want as a QA measure (ensuring that build dependencies are always 'known') or as a security measure (to protect against compiling potentially malicious code and/or running tests from such code)? The approaches in both cases are very different.

**Comment by jpakkane** [comment]
This memoization is only safe if the _entire_ directory tree is stored. Otherwise just changing any header in `/usr/include` breaks. There is also the requirement that any program executed (including generated executables) reliably produce the same output. 

What you are describing is very close to [reproducible builds](http://reproducible-builds.org/). We do support that (but this has not been properly verified and QA'd). Combining reproducible builds with a container such as Flatpak should provide what you need.

**Comment by nirbheek** [comment]
Yes, Flatpak uses [Project Bubblewrap](https://github.com/projectatomic/bubblewrap) which does exactly what you need. We could also add support for Bubblewrap, but we'd end up duplicating a lot of the functionality of Flatpak since it's a lot of work to create a new root with all the things that you need. 

[GNOME Builder](https://wiki.gnome.org/Apps/Builder) has native Flatpak integration. If you create a Meson project inside Builder, it will be built inside a Flatpak which gives you all those guarantees, and will also create a Flatpak app for you as a bonus (if you desire).

I think it also makes sense to add Flatpak support in Meson for building on CI and servers where Builder is not available. Just needs someone to do it. :)
